### DISCUSSION TITLE
Stop words parameter in completions API

### DISCUSSION THREAD

**Discussion by samuelbacaner** [moyix/fauxpilot#165] [Q&A] [answered]
The [example client curl](https://github.com/fauxpilot/fauxpilot/blob/main/documentation/client.md#curl-with-restful-apis) shows the `stop` param taking an array of strings and the `CodeGenProxy` transforms that into lists of tokens and offsets that Triton and FasterTrasnformer understand.

I'm trying to understand how the tokens and offsets are configured. [This method](https://github.com/fauxpilot/fauxpilot/blob/64c478d8261867c7a6d679a5632a7c49663b30ce/copilot_proxy/utils/codegen.py#L43-L73) iterates over each string encoding it into a list of token ids and then computes the offsets. In particular, I'm trying to understand the offsets in more detail and why offsets are padded with `-1` in [this line](https://github.com/fauxpilot/fauxpilot/blob/64c478d8261867c7a6d679a5632a7c49663b30ce/copilot_proxy/utils/codegen.py#L71).

Thanks!

**Comment by moyix** [comment, marked as answer]
It's a bit confusing, yes. Basically, FasterTransformer expects a tensor with shape [batch_size, 2, stop_words_length]. The first dimension is the batch dimension (so that each prompt in the batch can have its own unique list of stop words). The second dimension consists of the token IDs of the stop words (element 0) and the offsets of the boundaries between stop words (element 1). Finally, the last dimension is the tokens/offsets themselves.

The reason offsets and ids are padded is because the lists for each batch may not be equal, but tensors can't have ragged dimensions – so padding is needed. This line:

```python
pad_to = max(1, max(len(ids) for ids in flat_ids))
```

gets the length of the longest list of IDs across all the batches (or 1 if all the ids are empty lists). Moreover (again because they're all going into one big tensor) the ids and offsets also have to have the same length. But since len(offsets) ≤ len(ids) we only need to take the max over len(ids) rather than both.

The offsets themselves are pretty simple. The offset at position `i` gives the index of the *end* of word `i` in the token list. So for `['one', 'abracadabra']` we have:

```
ids = [505, 397, 11510, 324, 397, 430]
offsets = [1, 6]
```

So from the offsets list we see that the first word is `ids[0:1]` and the second word is `ids[1:6]`. The offsets are calculated by making a list of the length of each stop word in tokens and then calling cumsum() on that list.

Finally, why use -1 as padding for the offsets? Well, because it's guaranteed not to be a valid offset, so when decoding the list you can use it as a sentinel value to tell you where the list of tokens ends (even if it's padded).

As a full example:

```python
to_word_list_format([['one', 'abracadabra'], ['two']], tokenizer)
```

Returns:

```python
stop_words = \
array([[[  505,   397, 11510,   324,   397,   430],
        [    1,     6,    -1,    -1,    -1,    -1]],

       [[11545,     0,     0,     0,     0,     0],
        [    1,    -1,    -1,    -1,    -1,    -1]]], dtype=int32)
```

Decoding that, we start by going through `stop_words[0]` (the wordlist for the first prompt in our batch). `stop_words[0][1]` gives the offsets. `stop_words[0][1][0]` is 1 so our first word is `stop_words[0][0][0:1] == [505] # ("one")`, `stop_words[0][1][1]` is 6 so our second word is `stop_words[0][0][1:6] == [397, 11510,   324,   397,   430] # ("abracadabra")`, and `stop_words[0][1][2]` is -1 so we're done with the first wordlist.

In the second wordlist, `stop_words[1]`, we have `stop_words[1][1][0]` is 1, so the first word is `stop_words[1][0][0:1] == [11545] # ("two")`, and `stop_words[1][1][0]` is -1, so we're done (and we can ignore the padding 0s that make up the remainder of `stop_words[1][0]`.

**Comment by samuelbacaner** [comment]
Thanks for the detailed response, @moyix ! I really appreciate you taking the time to write this out. Makes perfect sense now.
